AG-304 Feature(ag-grid): dateTime and dateTimeString types

sdwvit opened the pull request:
This pull request introduces support for the 'dateTime' and 'dateTimeString' cell data types in AG-Grid:

### **Column Definitions**
- Updated column definitions to recognize 'dateTime' and 'dateTimeString'.

### **Filtering**
   - Added new filter options for 'dateTime' and 'dateTimeString' data types across components.
   - Enhanced tree list filtering to handle date-time values.

### **Code changes**
   - Added single-source-of-truth methods within the `DataTypeService` for 'dateTime' and 'dateTimeString'.

### **UI Components**
   - Upgraded filters, pickers, and other components to support time-aware data types.

### **Documentation Updates**
   - Added new data types to the documentation.
   - Provided examples demonstrating their usage in column definitions.

![image](https://github.com/user-attachments/assets/d8bf3c9e-6887-4827-9068-c18647ca1cee)
   
### **Enterprise-Specific Changes**:
   - Updated advanced filters to include new operators and display value models for time-aware data types.

Fixes [AG-14909](https://ag-grid.atlassian.net/browse/AG-14909) and maybe [AG-304](https://ag-grid.atlassian.net/browse/AG-304)

Test: https://plnkr.co/edit/7LbCplSa3miT463H?open=main.ts&preview and set date filter https://plnkr.co/edit/uMQYuVlEWIeP8hPR?open=main.ts&preview

sdwvit committed:
- Added 'dateTime' and 'dateTimeString' to BaseCellDataType
- Introduced DateTimeDataTypeDefinition and DateTimeStringDataTypeDefinition
- Updated BaseCellDataTypeDefMap to include dateTime options
- Defined ValueOf type for BaseCellDataTypeDefMap
- Modified DataTypeDefinition and CoreDataTypeDefinition types

sdwvit committed:
- Update return type in getDefaultDataTypes method

sdwvit committed:
fix types

sdwvit committed:
Implements filtering and data type operations for date and dateTime fields

sdwvit committed:
lint

sdwvit committed:
add dateTimeString to a priority list

sdwvit committed:
decoupled date/dateString and dateTime/dateTimeString

sdwvit committed:
lint

sdwvit committed:
- Updated `dataTypeMatchers` to have specific `BaseCellDataType` key types.
- Refactored assignment of `newDataTypeMatchers` with strong `BaseCellDataType` definition.
- Refactored `sortKeysInMatchers` by extracting it into a class method.

sdwvit committed:
- Updated defaultDateTimeFormatMatcher to correct time format regex

sdwvit committed:
- Extracted date and time format handling logic into a regex constant
- Updated logic for determining default filter based on data types
- Simplified switch case logic for default floating filter types

sdwvit committed:
- Refactored dateTime to date in generateExpressionOperators function
- Updated date and dateString operators with new dateOperatorsParams
- Added dateTime and dateTimeString operators with dateOperatorsParams

Commit:
wip: Expanding pre-defined cell data types to include 'dateTime' and 'dateTimeString'

Commit:
- Simplified the structure by defining generic methods for handling dates

Commit:
- move date regexp to date.ts

Commit:
- Changed the initialization logic to setTemplate method with updated parameters
- Added cellDataType to the eInput object for better data handling in editing cells

Commit:
- Refactored dateTimeRegexp as getDateTimeRegexp with optional time
- Added delimiter argument to _parseDateTimeFromString for custom parsing

Commit:
- Changed date/time related methods to use 'getDateTimeRegexp'
- Updated DATETIME_FIELD_TYPES array to include 'dateTime' and 'dateTimeString'
- Improved handling of date/time flags with new 'getDateIncludesTimeFlag' method

Commit:
- Modified DateCellEditor and DateCellEditorInput to inject DataTypeService
- Updated serialization logic in DateCellEditorInput and DateStringCellEditorInput
- Updated AgInputDateField constructor to handle includeTime flag
- Modified setMin and setMax methods in AgInputDateField to consider includeTime
- Updated parseDateTimeFromString call in AgInputDateField

Commit:
add todo for tomorrow

sdwvit committed:
- Using object to enforce development to implement new types as they are added
- Refactor DefaultDateComponent to dynamically set input type based on field types

sdwvit committed:
- Change date serialization to allow for custom separators
- Update date formatting logic to use specified separators

sdwvit committed:
update tests to comply with iso datetime format

sdwvit committed:
- Added a helper function to check if a string is a valid datetime
- Added a helper function to check if a string is a valid date (no time part)
- For that modified regexp to produce all delimiters and datetime parts
- _parseDateTimeFromString is now agnostic of delimiters and supports whatever Date() constructor supports

sdwvit committed:
- Update _parseDateTimeFromString to omit delimiter argument

sdwvit committed:
- Remove unnecessary delimiter arg in _parseDateTimeFromString
- Simplify _serialiseDate method arguments
- Decouple dateString and dateTimeString dataTypeMatchers

sdwvit committed:
- Update comments to clarify reasoning for prioritization

sdwvit committed:
- In DefaultDateComponent, set 'inputElement.type' based on conditions
- In AgAbstractInputField, added 'setExtraAttributes' method
- In AgInputDateField, dynamically set 'step' attribute for time inputs
- In AgInputTextField, support setting extra attributes dynamically

sdwvit committed:
- Better type enforcement

sdwvit committed:
- Explicit > Implicit

sdwvit committed:
- Remove 'extraAttributes' parameter from AgAbstractInputField constructor
- Update it's usage in AgInputTextField and AgInputDateField to reflect constructor changes

sdwvit committed:
- Refactor getDateObjectTypeDef and add getDateStringTypeDef methods

sdwvit committed:
- Rename colDefProps records to better suit their use
- Extracted dynamic records creation into class-instance properties, which aren't getting regenerated every call

sdwvit committed:
- Add 'includeTime' boolean property to AgInputDateFieldParams
- Modify AgInputDateField constructor to use 'includeTime' config option

sdwvit committed:
- Simplify _serialiseDate signature
- Remove unused dateTimeSeparator parameter in favor of default ' ' space
- Simplify DATE_TIME_REGEXP pattern
- Modify isValidDateTime logic for valid date with time parts

sdwvit committed:
- refactor getDefaultFilter functions a bit to enforce types + deduplicate code

sdwvit committed:
- revert some changes

sdwvit committed:
- simplify types a tiny bit

sdwvit committed:
- revert some changes & simplify types

sdwvit committed:
- revert some changes

sdwvit committed:
Merge remote-tracking branch 'origin/sdwvit/feature/ag-grid/column-filters-allow-date-input-filter-to-support-entering-filtering-by-time-values/AG-304' into sdwvit/feature/ag-grid/column-filters-allow-date-input-filter-to-support-entering-filtering-by-time-values/AG-304

sdwvit committed:
- revert some changes around date parsing
- enforce types and avoid dynamic call in advancedFilterBuilderItemComp.ts by replacing switch-case with a semi-static record.

sdwvit committed:
- revert some changes around date.ts

sdwvit committed:
- revert some changes around advancedFilterModel.ts

sdwvit committed:
- update Docs

sdwvit committed:
- update Docs

sdwvit committed:
- Adjust docs example
- Reorganized and updated type definitions in DataTypeService and entities

sdwvit committed:
- update docs
- fix a bug with custom matchers

sdwvit committed:
- update docs

sdwvit committed:
- update public exports

sdwvit committed:
wip

Commit:
- Added type definitions for advanced filter models and parameters in `advancedFilterModel.ts`
- Extracted `getDateParts` function in `date.ts` for reusability - that fixed a bug with Dates with time and Set filters
- Improved type safety and filter configuration mapping for different data types

Commit:
- Clean up type assertion in filter parameters

Commit:
Merge branch 'refs/heads/latest' into sdwvit/feature/ag-grid/column-filters-allow-date-input-filter-to-support-entering-filtering-by-time-values/AG-304

Commit:
lint

Commit:
- Added export for _getDateParts to community bundle

peterjrreynolds commented on documentation/ag-grid-docs/src/content/docs/cell-data-types/_examples/enable-cell-data-types/main.ts:
```diff
@@ -75,14 +79,22 @@ document.addEventListener('DOMContentLoaded', () => {
-                            parseInt(dateParts[0])
-                        ),
+                        date: dateString,
+                        dateObject: dateTimeObj,
```
`dateObject` needs to be populated with a `Date` without any time, otherwise the date filter won't work.

sdwvit replied:
well spotted, mind if I mention it in the docs? with the new date types, I foresee it may be a confusion for users. I can add a line about this behavior in this PR

peterjrreynolds commented on documentation/ag-grid-docs/src/content/docs/cell-data-types/_examples/pre-defined-cell-data-types/main.ts:
```diff
@@ -92,14 +96,22 @@ document.addEventListener('DOMContentLoaded', () => {
-                            parseInt(dateParts[0])
-                        ),
+                        date: dateString,
+                        dateObject: dateTimeObj,
```
Same as the other example

peterjrreynolds commented on documentation/ag-grid-docs/src/content/docs/cell-data-types/index.mdoc:
```diff
@@ -26,26 +26,32 @@ const gridOptions = {
 * The **Gold** column has a `'boolean'` data type.
 * The **Date** column has a `'date'` data type (cell values are `Date` objects).
 * The **Date (String)** column has a `'dateString'` data type (cell values are `string`s representing dates).
+* The **DateTime** column has a `'dateTime'` data type (cell values are `Date` objects, [and developer has to explicitly set `cellDataType: 'dateTime'`](#enable-cell-data-types), otherwise data type is set to just `'date'`).
```
```suggestion
* The **DateTime** column has a `'dateTime'` data type (cell values are `Date` objects). This is explicitly set to `cellDataType: 'dateTime'` as `Date` objects are always inferred to be `'date'` data type.
```
The docs assume the reader is a developer, so we wouldn't refer to them directly. Also, link text should be simple.

peterjrreynolds commented on documentation/ag-grid-docs/src/content/docs/cell-data-types/index.mdoc:
```diff
@@ -26,26 +26,32 @@ const gridOptions = {
 For inference to work for a column, it must contain non-null values and have the `field` property set. The resolved column definition (including the default column definition and column types) must also not have the Value Getter, Value Parser or reference data properties set, or be using [Sparklines](./sparklines-overview/). If these conditions are not met, no cell data type will be set (it will  ...
 
+{% note %}
+Because `'dateTime'` corresponds to cell values that are `Date` objects, there is no easy way to tell them apart from regular `'date'` columns. If you wish to enable `'dateTime'`'s higher precision fields, please [explicitly specify `cellDataType: 'dateTime'`](#enable-cell-data-types) in the corresponding `columnDefs` entry. 
```
```suggestion
Because `'dateTime'` corresponds to cell values that are `Date` objects, there is no easy way to tell them apart from regular `'date'` columns. If you wish to enable `'dateTime'`'s higher precision fields, please explicitly specify `cellDataType: 'dateTime'` in the corresponding `columnDefs` entry. 
```
Simple link text (don't really need the link here)

peterjrreynolds commented on documentation/ag-grid-docs/src/content/docs/cell-data-types/index.mdoc:
```diff
@@ -111,7 +117,31 @@ The following properties are set:
+
+The `'dateTime'` cell data type is used for date and time values that are represented as `Date` objects. Unlike the `'date'` cell data type which only shows the date portion, `'dateTime'` displays both date and time components.
+
+This data type uses the string format `'YYYY-MM-DD hh:mm:ss'`. If you wish to use a different format, you can [Override the Pre-Defined Cell Data Type Definition](#overriding-the-pre-defined-cell-data-type-definitions).
```
Are we not using the ISO format?

sdwvit replied:
ISO format is YYYY-MM-DD**T**hh:mm:ss**Z** (2025-05-23T14:26:58Z)

peterjrreynolds commented on documentation/ag-grid-docs/src/content/docs/cell-data-types/index.mdoc:
```diff
@@ -231,8 +261,8 @@ The following example demonstrates overriding pre-defined cell data types:
 {% gridExampleRunner title="Overriding Pre-Defined Cell Data Types" name="overriding-pre-defined-cell-data-types" /%}
 
-### Date as String Data Type Definition
+### Specific case for Date and DateTime as String Data Type Definition
```
```suggestion
### Date and DateTime as String Data Type Definition
```
Keep headings short where possible as the right-hand nav gets messy otherwise

peterjrreynolds commented on documentation/ag-grid-docs/src/content/docs/filter-advanced/_examples/advanced-filter/main.ts:
```diff
@@ -64,14 +68,22 @@ document.addEventListener('DOMContentLoaded', () => {
-                            parseInt(dateParts[0])
-                        ),
+                        date: dateString,
+                        dateObject: dateTimeObj,
```
See previous comment

peterjrreynolds commented on packages/ag-grid-community/src/columns/dataTypeService.ts:
```diff
@@ -510,17 +545,71 @@ export class DataTypeService extends BeanStub implements NamedBean {
+            colId,
+            formatValue,
+        });
+        Object.assign(colDef, partialColDef); // todo prefer object copy instead of mutation
```
Can remove todo.

We create col defs by applying in turn:
1. default col def
2. data type props
3. column types
4. col def

If we have a lot of columns (e.g. pivoting), and we created a copy each time, this would be expensive.

sdwvit replied:
Good insight, thank you

peterjrreynolds commented on packages/ag-grid-community/src/entities/dataType.ts:
```diff
@@ -109,32 +121,50 @@ export interface BooleanDataTypeDefinition<TData = any> extends BaseDataTypeDefi
 
 /** Represents a `'dateString'` data type (type `string` that represents a date). */
-export interface DateStringDataTypeDefinition<TData = any> extends BaseDataTypeDefinition<'dateString', TData, string> {
+export interface DateStringDataTypeDefinition<TData = any, K extends BaseCellDataType = 'dateString'>
```
I think this interface should be renamed and made private, and then exposed as another interface, e.g.
```
export interface DateStringDataTypeDefinition<TData = any> extends BaseDateStringDataTypeDefinition<TData, 'dateString'> {}

interface BaseDateStringDataTypeDefinition <TData = any, K extends BaseCellDataType = 'dateString'>
...
```

Otherwise we're exposing the second generic type param to users, and they shouldn't need to know anything about it.

sdwvit replied:
I think this is that case when extra abstraction actually hurts. Refactored using simpler approach

peterjrreynolds commented on packages/ag-grid-community/src/entities/dataType.ts:
```diff
@@ -109,32 +121,50 @@ export interface BooleanDataTypeDefinition<TData = any> extends BaseDataTypeDefi
-    | DateDataTypeDefinition<TData>
-    | DateStringDataTypeDefinition<TData>
-    | ObjectDataTypeDefinition<TData, any>;
+export type DataTypeDefinition<TData = any, TValue = any> = ValueOf<_DataTypeDefMap<TData, TValue>>;
```
Note that if you expand `dataTypeDefinitions` on the docs API widget, you see all these types.

Stuff done for the purpose of type checking should be private, and we should just expose a simple union.

sdwvit replied:
Do I understand correctly that only exported stuff is visible in docs?

peterjrreynolds commented on packages/ag-grid-community/src/filter/columnFilterService.ts:
```diff
@@ -54,14 +57,14 @@ const MONTH_KEYS: (keyof typeof MONTH_LOCALE_TEXT)[] = [
         return 1;
     }
-    return parseFloat(a) - parseFloat(b);
+    return parseFloat(String(a)) - parseFloat(String(b));
```
What was the motivator for changing this?

This is only used for number data type, so `a` and `b` should actually only ever be numbers here.

It's in the comparator, so there is a performance impact to this change.

sdwvit replied:
This is done only for correct typings, thank you for spotting the performance implication. Will revert

peterjrreynolds commented on packages/ag-grid-community/src/filter/columnFilterService.ts:
```diff
@@ -456,40 +459,39 @@ export class ColumnFilterService extends BeanStub implements NamedBean {
-            }
-        }
-        return defaultFilter;
+    private defaultFilters: Record<BaseCellDataType | 'set', UserComponentName> = {
```
Move properties to the top of the class with the other properties

sdwvit replied:
These do not depend on class instance. Okay to move them outside of class? see updated code

peterjrreynolds commented on packages/ag-grid-community/src/filter/columnFilterService.ts:
```diff
@@ -456,40 +459,39 @@ export class ColumnFilterService extends BeanStub implements NamedBean {
+        dateTimeString: 'agDateColumnFilter',
+        number: 'agNumberColumnFilter',
+        object: 'agTextColumnFilter',
+        set: 'agSetColumnFilter',
```
Might be neater if you just kept this to data types, and left the set logic in the method

peterjrreynolds commented on packages/ag-grid-community/src/filter/columnFilterService.ts:
```diff
@@ -886,148 +888,156 @@ export class ColumnFilterService extends BeanStub implements NamedBean {
+                colDef.filterParams =
+                    typeof colDef.filterParams === 'object'
+                        ? { colDef, ...colDef.filterParams, ...filterParamsOrNothing }
+                        : { colDef, ...filterParamsOrNothing };
```
Shouldn't be setting the `colDef` in the filter params here. This can just mimic the previous logic.

peterjrreynolds replied:
The issue here is actually that `DateFloatingFilter.getDateComponentParams` should be passing `filterParams` instead of `column.getColDef().filterParams` as the `filterParams` (the former includes all the grid-provided params like colDef, whereas the latter only includes the user-provided params).

peterjrreynolds commented on packages/ag-grid-community/src/filter/columnFilterService.ts:
```diff
@@ -886,148 +888,156 @@ export class ColumnFilterService extends BeanStub implements NamedBean {
+                        ? { colDef, ...colDef.filterParams, ...filterParamsOrNothing }
+                        : { colDef, ...filterParamsOrNothing };
+            } else {
+                colDef.filterValueGetter = filterParamsOrNothing;
```
I'm not a fan of this approach

sdwvit replied:
Elaborate please? This is [replicating the previous switch/case logic](https://github.com/ag-grid/ag-grid/pull/10542/files#diff-88ca7f3951a74023a6ecee154d8a469291c97c7c7ef0a6fe7c9291090f40c6e4L1015-L1029). The "not fun" part I can spot is the reliance on typeof filterParams === 'object' to decide strategy: it doesn't look clean to me neither, and enables future maintainer to make a mistake (hard to do, but still). 

If you confirm, I can explore ways to get rid of it.

peterjrreynolds commented on packages/ag-grid-community/src/interfaces/advancedFilterModel.ts:
```diff
@@ -98,8 +97,60 @@ export interface ObjectAdvancedFilterModel {
+}>;
+
+/** Represents a single filter condition on a column */
+export type ColumnAdvancedFilterModel = FilterModelDefMap[keyof FilterModelDefMap];
```
Similar issue to the data type definitions

peterjrreynolds commented on packages/ag-grid-community/src/utils/date.ts:
```diff
@@ -29,6 +29,28 @@ export function _serialiseDate(date: Date | null, includeTime = true, separator
+        return parts;
+    }
+
+    return parts.slice(0, 3);
```
Note that dates are more commonly used than datetimes, and creating intermediate arrays can have a performance impact.

sdwvit replied:
Gotcha, thank you for spotting this. Will refactor.

sdwvit committed:
- dates are more commonly used than datetimes, and creating intermediate arrays can have a performance impact, thus only stringify time if requested

sdwvit committed:
Update documentation/ag-grid-docs/src/content/docs/cell-data-types/index.mdoc

Co-authored-by: Pete Reynolds <40919976+peterjrreynolds@users.noreply.github.com>

sdwvit committed:
- dateObject needs to be populated with a Date without any time, otherwise the date filter won't work.

sdwvit committed:
Merge remote-tracking branch 'origin/sdwvit/feature/ag-grid/column-filters-allow-date-input-filter-to-support-entering-filtering-by-time-values/AG-304' into sdwvit/feature/ag-grid/column-filters-allow-date-input-filter-to-support-entering-filtering-by-time-values/AG-304

sdwvit committed:
Update documentation/ag-grid-docs/src/content/docs/cell-data-types/index.mdoc

Co-authored-by: Pete Reynolds <40919976+peterjrreynolds@users.noreply.github.com>

sdwvit committed:
- Update documentation regarding 'date' cell data type behavior

sdwvit committed:
Update documentation/ag-grid-docs/src/content/docs/cell-data-types/index.mdoc

Co-authored-by: Pete Reynolds <40919976+peterjrreynolds@users.noreply.github.com>

sdwvit committed:
- update docs example
- remove todo
- moved static properties outside of class

sdwvit committed:
Merge remote-tracking branch 'origin/sdwvit/feature/ag-grid/column-filters-allow-date-input-filter-to-support-entering-filtering-by-time-values/AG-304' into sdwvit/feature/ag-grid/column-filters-allow-date-input-filter-to-support-entering-filtering-by-time-values/AG-304

sdwvit committed:
- cast types in build time instead of the runtime

sdwvit committed:
address PR comments:
- pass base filterParams alongside with user defined filter params in floating filter for dates
- simplify data type interfaces
- refactor columnFilterService.ts a bit for readability and easier cognitive load
- docs update to mentioned time portion of a `'date'` column

sdwvit committed:
- simplify types

sdwvit commented on packages/ag-grid-community/src/entities/dataType.ts:
```diff
@@ -116,25 +128,47 @@ export interface DateStringDataTypeDefinition<TData = any> extends BaseDataTypeD
+export type CoreDataTypeDefinition<TData = any, TValue = any> = Omit<
+    DataTypeDefinition<TData, TValue>,
+    'extendsDataType'
+>;
```
@peterjrreynolds let me know if this is fine, or should I also fall back to union

peterjrreynolds replied:
This is only used internally, so fine

sdwvit committed:
- change date string parsing format, update tests, and docs
- update types

sdwvit committed:
- split set and regular filter params

sdwvit committed:
- add some tests and fix isValidDateTime behavior

sdwvit committed:
- lint
- address PR comment: keep set filter logic in the method

sdwvit committed:
- less frequent calculations

sdwvit committed:
fix bugs

Commit:
- update tests

sdwvit committed:
- Update note

sdwvit committed:
- address PM spec comments
